### ISSUE TITLE
Bug in alt! macro

### ISSUE THREAD

**Issue by Selat** [Geal/nom#133]
I think I've found a problem with the alt! macro. Consider the following code:

```rust
#[macro_use]
extern crate nom;

use nom::{digit, multispace, IResult};
use std::str;

named!(pub token<&[u8], Vec<&str> >,
       many1!(chain!(v:alt!(
           tag!("t1") => {|_| "t1"} |
           delimited!(opt!(multispace),
                      chain!(tag!("."),
                             || "t2"),
                      opt!(multispace)) |
           map_res!(delimited!(opt!(multispace), digit, opt!(multispace)),
                    str::from_utf8)),
                     || v)));

fn main() {
    let mut buffer : [u8; 3] = [0; 3];
    buffer[0] = '1' as u8;
    buffer[1] = ' ' as u8;
    buffer[2] = '2' as u8;
    let s = token(&buffer);
    match s {
        IResult::Done(_, o) => {
            println!("Length: {}", o.len());},
        _ => println!("Failed to parse!"),
    }
}
```

It outputs
```
Length: 1
```

But if you comment out the first pattern from the alt! macro
```rust
named!(pub token<&[u8], Vec<&str> >,
       many1!(chain!(v:alt!(
           //tag!("t1") => {|_| "t1"} |
           delimited!(opt!(multispace),
                      chain!(tag!("."),
                             || "t2"),
                      opt!(multispace)) |
           map_res!(delimited!(opt!(multispace), digit, opt!(multispace)),
                    str::from_utf8)),
                     || v)));
```
The output will be
```
Length: 2
```

As the pattern tag!("t1") doesn't match anything in both cases, the output should be the same :smiley:

**Comment by KeenS** [comment]
I know another example

```rust
#[macro_use]
extern crate nom;

use nom::{digit, multispace, IResult};
use std::str;

named!(pub token<&[u8], Vec<&str> >,
       many1!(chain!(v:alt!(
           map_res!(delimited!(opt!(multispace), digit, opt!(multispace)),
                    str::from_utf8)|
           tag!("t1") => {|_| "t1"} |
           delimited!(opt!(multispace),
                      chain!(tag!("."),
                             || "t2"),
                      opt!(multispace))
 ),
                     || v)));

fn main() {
    let mut buffer : [u8; 3] = [0; 3];
    buffer[0] = '1' as u8;
    buffer[1] = ' ' as u8;
    buffer[2] = '2' as u8;
    let s = token(&buffer);
    match s {
        IResult::Done(_, o) => {
            println!("Length: {}", o.len());},
        _ => println!("Failed to parse!"),
    }
}
```

```
Length: 2
```

I changed only the submacro order in `alt`. It seems that the second submacro performs oddly

**Comment by Geal** [comment]
Thank you for your report!

It appears the bug is not in `alt!` but in `many1!`:

With the input you provide, the first application of `alt!` receives "1 2", parses "1", and leaves "2" as second input. If the `tag!("t1")` is not there, everything works correctly and the result of `token` is `["1", "2"]`.
If `tag!("t1")` is there, it receives "2" as input, and since "2" is smaller than the tag it tries to recognize, it returns `Incomplete(Size(2))`. Since one of the branches of `alt!` returned `Incomplete`, it does the same (this is an expected behaviour). But `many1` has a bug: if one of its parser applications returned `Incomplete`, it stops and returns `Done`.

You can see it more easily with this version of the code without `many1!`, applying the parser multiple times manually (by the way, you could remove `chain!` there, you did not need it):

```rust
#[macro_use]
extern crate nom;

use nom::{digit, multispace, IResult};
use std::str;

//named!(pub token<&[u8], Vec<&str> >,
named!(pub token<&[u8], &str >,
       //many1!(
       //chain!(v:
         alt!(
           tag!("t1") => {|_| "t1"} |
           delimited!(opt!(multispace),
                      chain!(tag!("."),
                             || "t2"),
                      opt!(multispace)) |
           map_res!(delimited!(opt!(multispace), digit, opt!(multispace)),
                    str::from_utf8))
       //, 
       //|| v)
       //) 
);

fn main() {
    let mut buffer : [u8; 3] = [0; 3]; 
    buffer[0] = '1' as u8; 
    buffer[1] = ' ' as u8; 
    buffer[2] = '2' as u8; 
    let s = token(&buffer);
    match s { 
        IResult::Done(i, o) => {
            println!("Length: {}", o.len());
            println!("content: {:?}", o); 
            println!("second result: {:?}", token(i));
        },  
        r => {
          println!("Failed to parse!: {:?}", r)
        },  
    }   
}
```

So I'll fix the bug in `many0!` and `many1!`, but in the meantime, if you want to avoid the effects of incomplete data, you can wrap a combinator with `complete!`. This combinator wraps a parser, and if the parser returns `Incomplete`, `complete!` will return an `Error`. This will transform the tag line like this: `complete!(tag!("t1")) => {|_| "t1"} | `.

**Comment by KeenS** [comment]
Thank you for your investigation and suggestion of workaround. I don't have to wait for bugfix!

**Comment by Geal** [comment]
The fix is in master: 7666c7abcddd9da8e8ec7f98599cf2af1eb5d004. It will be available in 1.1. Also a case like the one you got will appear less, since `tag!` will now match on incomplete data (ie, if there's on char, and it is already different, return an error instead of incomplete).
